### ISSUE TITLE
Reproduction of Regression Results Fails (for me :-)  )

### ISSUE THREAD

**Issue by haimivan** [marcotcr/lime#189]
Hi,

I would like to try to reproduce the regression example. 
My assumption is that if I multiply the weights that I get from `exp.as_list` with the feature values and divide it by the amount of features, that I would get the predicted value. Unfortunately, I get a different result.

My code looks like the following:

    # coding: utf-8
    from sklearn.datasets import load_boston
    import sklearn.ensemble
    import numpy as np
    from sklearn.model_selection import train_test_split
    import lime
    import lime.lime_tabular
    import pandas as pd
    
    import numpy as np
    
    random_seed = np.random.seed(42)
    
    boston = load_boston()
    print(boston['feature_names'])
    
    rf = sklearn.ensemble.RandomForestRegressor(n_estimators=1000)
    train, test, labels_train, labels_test = train_test_split(boston.data, boston.target, train_size=0.80, random_state=random_seed)
    rf.fit(train, labels_train)
    
    print('Random Forest MSError', np.mean((rf.predict(test) - labels_test) ** 2))
    
    print('MSError when predicting the mean', np.mean((labels_train.mean() - labels_test) ** 2))
    
    categorical_features = np.argwhere(
        np.array([len(set(boston.data[:, x]))
                  for x in range(boston.data.shape[1])]) <= 10).flatten()
    
    for runner, x in enumerate(range(boston.data.shape[1])):
        print(x, np.array([len(set(boston.data[:, x]))]))
    
    explainer = lime.lime_tabular.LimeTabularExplainer(train,
                                                        random_state=42,
                                                       feature_names=boston.feature_names,
                                                       class_names=['price'],
                                                       categorical_features=categorical_features,
                                                       verbose=True, mode='regression')
    
    i = 100
    data_to_explain = test[i]
    
    exp = explainer.explain_instance(data_to_explain, rf.predict)
    exp.show_in_notebook(show_table=True)
    import webbrowser
    
    html = exp.as_html(show_table=True)
    with open("tmp_file.html", "w", encoding="utf-8") as text_file:
        text_file.write(html)
    webbrowser.open_new_tab("tmp_file.html")
    
    print("exp.as_list():")
    for runner in exp.as_list():
        print(runner[0], "\t", runner[1])
    
I take the result at the end:

    exp.as_list():
    5.89 < RM <= 6.21 	 -2.9728642977516193
    10.93 < LSTAT <= 16.37 	 -1.3551478944259272
    TAX <= 279.00 	 0.49615341040567384
    AGE <= 45.55 	 0.41870398678519766
    16.80 < PTRATIO <= 18.70 	 0.4184435325025001
    NOX <= 0.45 	 0.3617742436741252
    B > 395.75 	 0.3502449121041437
    INDUS <= 5.13 	 0.3150307076861154
    RAD=3 	 -0.3104442580102656
    CHAS=0 	 0.2796530015321413

and put it together with the feature values in Excel:

    | feature 	| feature-value 	| factor             	| factor*feature     	|   	|
    |---------	|---------------	|--------------------	|--------------------	|---	|
    | RM      	| 6.01          	| -2.97286429775162  	| -17.8669144294872  	|   	|
    | LSTAT   	| 12.86         	| -1.35514789442593  	| -17.4272019223174  	|   	|
    | TAX     	| 247           	| 0.496153410405674  	| 122.549892370201   	|   	|
    | AGE     	| 45.1          	| 0.418703986785198  	| 18.8835498040124   	|   	|
    | PTRATIO 	| 18.5          	| 0.4184435325025    	| 7.74120535129625   	|   	|
    | NOX     	| 0.45          	| 0.361774243674125  	| 0.162798409653356  	|   	|
    | B       	| 395.99        	| 0.350244912104144  	| 138.69348274412    	|   	|
    | INDUS   	| 4.49          	| 0.315030707686115  	| 1.41448787751066   	|   	|
    | RAD=3   	| TRUE          	| -0.310444258010266 	| -0.310444258010266 	|   	|
    | CHAS=0  	| TRUE          	| 0.279653001532141  	| 0.279653001532141  	|   	|
    |         	|               	| sum/10         	| 25.4120508948511   	|   	|

I would expect to get as a result the predicted value of 20.28, but not a value of 25.4120508948511.
(Substituting the "TRUE"s with "3" and "0", results in: 25.3219967430959, so still a different result).

Maybe I am understanding something wrong?

Thanks in advance!

**Comment by marcotcr** [comment]
You're not guaranteed in getting the exact predicted value, but it should be close.
You should also add the value of exp.intercept, that probably accounts for the difference.

**Comment by haimivan** [comment]
As the result of `exp = explainer.explain_instance(data_to_explain, rf.predict)` I get:

    Intercept 24.000430356485413
    Prediction_local [22.0019777]
    Right: 20.279300000000056

so _adding_ `Intercept` would lead in the wrong direction, wouldn't it?

**Comment by marcotcr** [comment]
Prediction_local is intercept + weights * features. 
Sorry, this is poorly documented, but the weights are actually relative to the scaled data, so your feature_value column should be scaled. You can get the scaled values by running:
`(data_to_explain - explainer.scaler.mean_) / explainer.scaler.scale_`
Or you can believe that I did it right when computing prediction_local : )

**Comment by haimivan** [comment]
Hi,

thanks for your explanation.

Unfortunately, the variables mentioned have the following values:

    explainer.scaler.mean_: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
    explainer.scaler.scale_: [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]

So does it mean that the scaled values are (always) the same as the original feature_values?

Thanks in advance!

**Comment by haimivan** [comment]
Hi @marcotcr ,

I would like to make a presentation on LIME in my company in some days.

Did you perhaps already had the chance to look at my last comment?

Thanks in advance!

**Comment by marcotcr** [comment]
Ah, it's because TabularExplainer discretizes continuous features by default (`discretize_continuous=True` in `init`). If you look at the features in the explanations, they are all in discrete slices (e.g. 5.89 < RM <= 6.21), so you should multiply their weight by 1.
If you add all of the weights in the explanations to the intercept, you get 22.0019777 (which is prediction_local)
